openshift-tools/pkg/discover-etcd-initial-cluster/initial-cluster.go

```diff
@@ -105,173 +109,197 @@ func (o *DiscoverEtcdInitialClusterOptions) Validate() error {
+		fmt.Fprintf(os.Stderr, "dataDir is present on %s\n", o.TargetName)
+		dataDirExists = true
+	}
 
 	client, err := o.getClient()
-	if err != nil && memberDirExists {
-		// we weren't able to get client and need to return based previous memberDir so we can restart.  This is the off and on again flow.
-		fmt.Fprintf(os.Stderr, "Couldn't get client, but memberDir %s is present on %s, err=%s. Returning with no error.\n", memberDir, o.TargetName, err)
-		fmt.Printf(o.TargetName)
+
+	// Condition: unknown membership but dataDir exists
+	// possible reasons for this condition.
+	// 1.) single node etcd cluster
+	// 2.) transient networking problem
+	// 3.) on and off flow
+	// Result: empty initial config
+	if err != nil && dataDirExists {
+		fmt.Fprintf(os.Stderr, "failed to create etcd client, but the server is already initialized as member %q before, starting as etcd member: %v", o.TargetName, err.Error())
 		return nil
-	} else if err != nil {
-		return err
+	}
+	if err != nil {
+		return fmt.Errorf("failed to create etcd client: %v",err)
 	}
 	defer client.Close()
 
-	var targetMember *etcdserverpb.Member
-	var allMembers []*etcdserverpb.Member
+	target := url.URL{
+		Scheme: o.TargetPeerURLScheme,
+		Host:   fmt.Sprintf("%s:%s", o.TargetPeerURLHost, o.TargetPeerURLPort),
+	}
+
 	for i := 0; i < 10; i++ {
 		fmt.Fprintf(os.Stderr, "#### attempt %d\n", i)
-		targetMember, allMembers, err = o.checkForTarget(client)
 
-		for _, member := range allMembers {
-			fmt.Fprintf(os.Stderr, "      member=%v\n", stringifyMember(member))
+		// check member list on each iteration for changes
+		cluster, err := client.MemberList(context.TODO())
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "member list request failed: %v", err)
+			continue
+		}
+		logCurrentMembership(cluster.Members)
+		targetMember, found := checkTargetMember(target, cluster.Members)
+
+		// Condition: unstarted member found
+		// This member is part of the cluster but has not yet started. We know this because the name is populated at
+		// runtime which this member does not have.
+		// Result: provide initial cluster value so etcd can communicate with peers during startup.
+		if found && targetMember.Name == "" && !dataDirExists {
+			fmt.Print(getInitialCluster(o.TargetName, targetMember, cluster.Members))
+			return nil
 		}
-		fmt.Fprintf(os.Stderr, "      target=%v, err=%v\n", stringifyMember(targetMember), err)
 
-		// we're done because we found what we want.
-		if targetMember != nil && err == nil {
-			break
+		// Condition: unstarted member found with dataDir
+		// This member is part of the cluster but has not yet started yet has a dataDir
+		// Result: archive old dataDir and return Error
+		if found && targetMember.Name == "" && dataDirExists {
+			archivedDir, err := archiveDataDir(o.DataDir)
+			if err != nil {
+				return err
+			}
+			return fmt.Errorf("member %q is unstarted but previous members dataDir exists: archiving to %q", target.String(), archivedDir)
```

Comment:
The multiple conditions on these if statements may make it confusing to understand the logic. In fact, I think this may be a case where nested if statements may make it easier to understand the logic than multiple conditions on a single line. 

Here I drew the flowchart for explaining the logic of discovering initial cluster:
https://lucid.app/lucidchart/invitations/accept/584c5c49-3fad-4531-be0c-cf111b360c6b?viewport_loc=283%2C104%2C2208%2C2461%2C0_0

Reply:
Should we return error, or success and initial string on line 184?

Reply:
My thoughts here are that we have taken action on an existing member. We should return an error and have the container start again in which case scale-up will be a success. It feels technicaly that etcd is trying to start in an unexpected condition. As a result we clear that condition report error message then restarts the container where it can go through the logical checks again.

Reply:
> Here I drew the flowchart for explaining the logic of discovering initial cluster:

Looks good! I think the flowchart would be a great addition to docs[1] if you would like to raise that PR.

In general, I am not a fan of nested conditions if it is not required. As far as readability having the conditions on the first level means the logical gate is human readable on one line thus easier to understand IMO. The comment `Condition` above summarizes the logic as well?

```
if found && targetMember.Name == "" && dataDirExists {
```

"if the member is found with no name and dataDir exists do something" 

As a followup I will update the comments to align with the tested scenarios[2] 

[1] https://github.com/openshift/etcd/pull/73/files#diff-3d1414e0d47047e4fdcae958fb7654d082c1eb1b71cf4eb38aeb3190db678208R61

[2] https://github.com/openshift/etcd/pull/73/files#diff-36ef21de0b9a0df2eaebbfcb7ae9c062a7c06921f4b71af73a472ec4b4a493a0R95